### PR TITLE
Metal3 remediation backport

### PR DESCRIPTION
**Pull request by slintes**
https://issues.redhat.com/browse/ECOPROJECT-92
https://issues.redhat.com/browse/METAL-224

This is the first of two PR for backporting metal3 remediation.
Part 2: https://github.com/openshift/cluster-api-provider-baremetal/pull/184

What I've done here in a nutshell is adding webhooks for metal3 remediation:

- add new service manifest to /install because the machine container served no webhook yet
- sync.go: add certificate mount and port config to machine-controllers pod / machine container
- webhooks/config.go: move all existing config code here, and add remediation config

### CODE REVIEW

File: pkg/webhooks/config.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+			Name: defaultWebhookConfigurationName,
+			Annotations: map[string]string{
+				"service.beta.openshift.io/inject-cabundle": "true",
+			},
+		},
+		Webhooks: []admissionregistrationv1.ValidatingWebhook{
+			MachineValidatingWebhook(),
+			MachineSetValidatingWebhook(),
+			Metal3RemediationValidatingWebhook(),
+			Metal3RemediationTemplateValidatingWebhook(),
```
Comment (JoelSpeed) [inline pkg/webhooks/config.go:51]: Do we want to deploy this on all platforms? Every machine action will check for a webhook that is unlikely to be there in more cases than it is likely to be there.

Perhaps these should be applied as a separate `ValidatingWebhookConfiguration` rather than joined with the other Machine API ones?

Comment (slintes) [reply pkg/webhooks/config.go:51]: Maybe I misunderstand your concern, but I don't think machine actions will check these new webhooks. Part of the webhook configuation is the kind for which they are registered 🤔 E.g.
```
				Rule: admissionregistrationv1.Rule{
					APIGroups:   []string{capm3apis.GroupVersion.Group},
					APIVersions: []string{capm3apis.GroupVersion.Version},
					Resources:   []string{"metal3remediations"},
				},
```

Comment (JoelSpeed) [reply pkg/webhooks/config.go:51]: Right ok, I thought we were adding new machine webhooks which would have been more impactful.

I am still concerned we are setting a precedent though, would it not be better for each subsystem to maintain their own MWC/VWC rather than just adding more and more config to the existing Machine one?

Comment (slintes) [reply pkg/webhooks/config.go:51]: What do you mean with "adding more and more config to the existing Machine one"?

The new MWC / VWC are not related to Machines at all. They are independent MWC / VMC resources. That's actually why I moved  all configs into `config.go`, instead of adding the new ones to `machine_webhook.go` (which was done for machineset configs before), to avoid confusion. See 1st commit.

The only thing which makes this related to machines is, that the container name of the provider is `machine-controller`, so some names of new ports/mounts also use `machine`, to make it clear to which container they belong.

Comment (JoelSpeed) [reply pkg/webhooks/config.go:51]: I mean that this Pr is adding additional `Webhooks` to an existing `ValidatingWebhookConfiguration`, so when I look at `oc get validatingwebhookconfigurations` I will see `machine-api` with 4 webhooks defined within it.

I think for separation of concerns it's likely better if the existing `machine-api` webhook configuration retains it's original 2 webhooks and we create a separate `machine-api-metal3-remediation` ValidatingWebhookConfiguration that handles the `Metal3Remediation...` webhooks

Comment (slintes) [reply pkg/webhooks/config.go]: Oh sorry, now I got it. Done!

File: pkg/operator/operator.go [hunk 2]
```diff
@@ -142,11 +144,13 @@ func New(
 	optr.proxyLister = proxyInformer.Lister()
 	optr.proxyListerSynced = proxyInformer.Informer().HasSynced
 
-	optr.validatingWebhookCache = resourceapply.NewResourceCache()
+	optr.validatingMachineWebhookCache = resourceapply.NewResourceCache()
+	optr.validatingMetal3RemediationWebhookCache = resourceapply.NewResourceCache()
```
Comment (JoelSpeed) [inline pkg/operator/operator.go:148]: I don't remember much about these caches, they can't be shared?

**Review by JoelSpeed — COMMENTED** [review]
OK I think this is pretty much ready to go. I would like to see another review from a MAPI maintainer before we merge, will get someone else to give this a look over

**Comment by JoelSpeed** [comment]
Could we ask QE to take a look at this and verify that this is working as expected pre-merge?

File: pkg/operator/operator.go (hunk 2 above)
Comment (slintes) [reply pkg/operator/operator.go:148]: Good question... I thought so, because both old configs had their own cache already.
But looking at the cache code I think all resources can use the same cache... 🤔

**Comment by lobziik** [comment]
Overall looks ok to me.

The only thing which i would appreciate to see: https://github.com/openshift/machine-api-operator/pull/1075/files#diff-f88e59a0d3c8f1f4b30a0e1c091ba2e1b08db33f18dd01fc4a9118062d9df8a0R293 can we expand this docstrings with some documentation/gr repo references? Just for give some breadcrumbs to future us about what is this actually

**Comment by slintes** [comment]
rebased while hoping to get a review on the PR...

**Comment by slintes** [comment]
wait for #1115 and/or #1100 to get in, it will help fixing the unit test at least (compile error due to version mismatch k8s/cluster-api/metal3)

**Comment by slintes** [comment]
> wait for #1115 and/or #1100 to get in, it will help fixing the unit test at least (compile error due to version mismatch k8s/cluster-api/metal3)

well, let's give new versions a try already... looking forward to merge later 😉

**Review by JoelSpeed — COMMENTED** [review]
In general I think this is ok, I'd like someone else to give it a review for LGTM though

File: pkg/operator/operator.go [hunk 3]
```diff
@@ -123,11 +121,11 @@ func New(
 		operandVersions: operandVersions,
 	}
 
-	deployInformer.Informer().AddEventHandler(optr.eventHandlerDeployments())
-	daemonsetInformer.Informer().AddEventHandler(optr.eventHandler())
-	validatingWebhookInformer.Informer().AddEventHandler(optr.eventHandlerSingleton(isMachineWebhook))
-	mutatingWebhookInformer.Informer().AddEventHandler(optr.eventHandlerSingleton(isMachineWebhook))
-	featureGateInformer.Informer().AddEventHandler(optr.eventHandler())
+	_, _ = deployInformer.Informer().AddEventHandler(optr.eventHandlerDeployments())
```
Comment (lobziik) [inline pkg/operator/operator.go:124]: Hm, I'm not quite sure why is this needed...

@slintes were there some issues with linter without this?

File: pkg/operator/sync_test.go [hunk 4]
```diff
@@ -7,6 +7,7 @@ import (
 	"time"
 
 	. "github.com/onsi/gomega"
+	v1 "github.com/openshift/api/config/v1"
```
Comment (lobziik) [inline pkg/operator/sync_test.go:10]: nit: could you please split these import blocks? :)

**Review by lobziik — COMMENTED** [review]
Overall lgtm, just a couple of nit comments.

@JoelSpeed feel free to move this on if we need it to be merged. No major concerns from me.

File: pkg/operator/operator.go (hunk 3 above)
Comment (slintes) [reply pkg/operator/operator.go:124]: yes indeed, linter was complaining about unevaluated retirn values

File: pkg/operator/sync_test.go (hunk 4 above)
Comment (slintes) [reply pkg/operator/sync_test.go:10]: sure, done

**Review by lobziik — COMMENTED** [review]
/lgtm

Thanks a lot!

**Comment by slintes** [comment]
@JoelSpeed @lobziik thanks a lot for your reviews, much appreciated 🙂 👍🏼
